[jit] fix inliner bug

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25089 [jit] Emit script function calls during tracing.
* **#25052 [jit] fix inliner bug**

Previously we would not inline nested functions, now we do.

Differential Revision: [D16973848](https://our.internmc.facebook.com/intern/diff/D16973848)

suo committed:
[jit] fix inliner bug

Previously we would not inline nested functions, now we do.

suo committed:
Update on "[jit] fix inliner bug"


Previously we would not inline nested functions, now we do.

Differential Revision: [D16973848](https://our.internmc.facebook.com/intern/diff/D16973848)

suo commented on torch/csrc/jit/passes/inliner.cpp:
```diff
@@ -12,23 +12,26 @@ using namespace ::c10::prim;
+        // and potentially replace it with new nodes.
+        it--;
         inlineCallTo(cur, *fun_type->function()->graph());
-        if (!function_constant->hasUses()) {
```
I removed this because it complicates the `it--` logic above. I think it's okay to litter these—we should be running DCE at some point anyway, I don't think needs to be a guarantee that function constants are cleaned up by the inliner. What do people think?

ZolotukhinM replied:
> I removed this because it complicates the `it--` logic above. I think it's okay to litter these—we should be running DCE at some point anyway, I don't think needs to be a guarantee that function constants are cleaned up by the inliner. What do people think?

You could save these constants into some `to_remove` list and wipe them all off in the end. But I agree that we would want to run cleanup passes after inlining anyway, so up to you.

Krovatkin replied:
@suo The historical reason for this logic is that I needed to make sure that after **Inliner's** inlining, graphs would be identical to the graphs after inlining in `compiler.cpp.` I couldn't run DCE since it would remove more than just those constants. Letting DCE handle this seems like the right thing to do.

ZolotukhinM commented on torch/csrc/jit/passes/inliner.cpp:
```diff
@@ -12,23 +12,26 @@ using namespace ::c10::prim;
         auto fun_type =
             function_constant->output()->type()->expect<FunctionType>();
         cur->removeInput(0);
+        // We need to decrerement `it` because inlineCallTo will destroy `cur`
```
Typo: decrerement

ZolotukhinM commented on torch/csrc/jit/passes/inliner.cpp:
```diff
@@ -12,23 +12,26 @@ using namespace ::c10::prim;
         const std::string& name = cur->s(attr::name);
         auto function =
             cur->input(0)->type()->expect<ClassType>()->getMethod(name);
+        // We need to decrerement `it` because inlineCallTo will destroy `cur`
```
Same typo :)

ZolotukhinM reviewed (approved):
👍

suo committed:
Update on "[jit] fix inliner bug"


Previously we would not inline nested functions, now we do.

Differential Revision: [D16973848](https://our.internmc.facebook.com/intern/diff/D16973848)

suo commented:
discussion with @zdevito and @ZolotukhinM offline: we probably don't want the inliner to be recursive by default, as it leads to potential n^2 behavior. What we should do is have `Function` hold a "generically optimized graph" (e.g. run all optimizations that don't require input specialization) and ask for that during inlining. That way we perform inlining for a callee at most once.

suo committed:
Update on "[jit] fix inliner bug"


Previously we would not inline nested functions, now we do.

Differential Revision: [D16973848](https://our.internmc.facebook.com/intern/diff/D16973848)

suo committed:
Update on "[jit] fix inliner bug"


Previously we would not inline nested functions, now we do.

Differential Revision: [D16973848](https://our.internmc.facebook.com/intern/diff/D16973848)

suo committed:
Update on "[jit] fix inliner bug"


Previously we would not inline nested functions, now we do.

Differential Revision: [D16973848](https://our.internmc.facebook.com/intern/diff/D16973848)
